src/data/message_formats.py

```diff
@@ -1,92 +1,111 @@
-    ) -> dict[str, str | list[dict[str, str]]]:
+        answer: Optional[str] = None,
+        key_object_info: Optional[dict] = None,
+        context: Optional[List[Tuple[str, str]]] = None,
+    ) -> Dict[str, Any]:
         pass
 
 
 class QwenMessageFormat(MessageFormat):
     def format(
         self,
         question: str,
-        key_object_info: dict,
         image_path: str,
         system_prompt: str,
-        answer: str="",
-    ) -> dict[str, str | list[dict[str, str]]]:
+        answer: Optional[str] = None,
+        key_object_info: Optional[dict] = None,
+        context: Optional[List[Tuple[str, str]]] = None,
+    ) -> Dict[str, Any]:
+        content = [
+            {"type": "text", "text": system_prompt},
+            {"type": "text", "text": "Question: " + question},
+            {
+                "type": "image",
+                "image": f"file://{image_path}",
+            },
+        ]
+
+        if key_object_info:
+            content.append(
+                {
+                    "type": "text",
+                    "text": "Key object infos:\n" + key_object_info.__str__(),
+                }
+            )
+
+        if context:
+            for context_q, context_a in context:
+                content.append(
+                    {"type": "text", "text": f"Context Question: {context_q}"}
+                )
+                content.append({"type": "text", "text": f"Context Answer: {context_a}"})
+
         return {
             "role": "user",
-            "content": [
-                {"type": "text", "text": system_prompt},
-                {"type": "text", "text": "Question: " + question},
-                {
-                    "type": "image",
-                    "image": f"file://{image_path}",
-                },
-                *(
-                    [
-                        {
-                            "type": "text",
-                            "text": "Key object infos:\n" + key_object_info.__str__(),
-                        }
-                    ]
-                    if key_object_info
-                    else []
-                ),
-            ],
+            "content": content,
         }
 
 
 class QwenTrainingMessageFormat(MessageFormat):
     """
     Message format adhering to OAI implementation
     """
+
     def format(
         self,
         question: str,
-        key_object_info: dict,
         image_path: str,
         system_prompt: str,
-        answer: str,
-    ) -> dict[str, str | list[dict[str, str]]]:
+        answer: Optional[str] = None,
+        key_object_info: Optional[dict] = None,
+        context: Optional[List[Tuple[str, str]]] = None,
+    ) -> Dict[str, Any]:
+        content = [
+            {"type": "text", "text": system_prompt},
+            {"type": "text", "text": "Question: " + question},
+            {
+                "type": "image",
+                "image": f"file://{image_path}",
+            },
+        ]
+
+        if key_object_info:
```

Comment:
if we pass the kois and the context dont we make either context or kois redundat?